### ISSUE TITLE
Missing `removeDependency` on the `InheritedElement`

### ISSUE THREAD

**Issue by techvx** [flutter/flutter#128432]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Use case

Preface: I am aware of @rrousselGit's previous #107112 PR and the reasons as to why it wasn't accepted. As many have outlined, merging the reactivity/listen-ability with inherent dependency removal is an inherently risky path to tread on, which can introduce more problems than it can solve down the line. 

That doesn't, however, make the absence of `removeDependency` to complete the current set of `update`, `get`, `set` and `notify` methods, any less tolerable. Not having a basic `remove` makes the whole set virtually useless for any cases more complex than trivial, as in using a `ValueNotifier` that only allows to `get`, `set`, and `notify` its listeners, but never to remove any of them.

Any logic on the side of `InheritedElement` which involves keeping track of dependencies separately, for any reason dictated by an appropriate use case, as it currently stands:

```dart
class CustomInheritedElement 
extends InheritedElement {

  CustomInheritedElement(super.widget);
  final Map<Tracked, Set<Element>> _tracked; 

  // get, set, update, and so on
}

```

Requires forceful iteration over the whole structure on each and every minor change to determine whether the tracked `Element`s are still properly `mounted` or not and can be safely notified via `didChangeDependencies()`, as needed.

### Proposal

Given the [previous](https://github.com/flutter/flutter/pull/107112#issuecomment-1185998599) comments and [no](https://github.com/flutter/flutter/pull/107112#issuecomment-1201891006) particular objections to the `remove` itself - [recognized](https://github.com/flutter/flutter/pull/107112#issuecomment-1201896460) as an issue by the core team, and [acknowledged](https://github.com/flutter/flutter/pull/107112#issuecomment-1201896460) as one that must be fixed, I'd like to get a green light on the potential merge of the `removeDependency` PR only, with no additional gimmicks and "nice-to-have"-s.

**Comment by rrousselGit** [comment]
> Given the https://github.com/flutter/flutter/pull/107112#issuecomment-1185998599 comments and https://github.com/flutter/flutter/pull/107112#issuecomment-1201891006 particular objections to the remove itself - https://github.com/flutter/flutter/pull/107112#issuecomment-1201896460 as an issue by the core team, and https://github.com/flutter/flutter/pull/107112#issuecomment-1201896460 as one that must be fixed, I'd like to get a green light on the potential merge of the removeDependency

The acknowledgment/recognition linked technically are not about `removeDependency` but  https://github.com/flutter/flutter/issues/106549

The PR I made is about two separate but related issues. 

Still, we could keep the changes specific to `removeDependency` and tackle https://github.com/flutter/flutter/issues/106549 separately

**Comment by rrousselGit** [comment]
I haven't forgotten that I still need to write a design doc to solve those issues btw. I'm just busy, unfortunately.

About this @goderbauer, if I want to try various implementations to solve https://github.com/flutter/flutter/issues/106549, are there good existing benchmarks I could follow to see the performance impact?  
After-all, ultimately the main issue seems to be performance

**Comment by techvx** [comment]
Performance considerations aren't of as much of a concern in this case as the hard-coded current `deactivate()` logic in the `Element`-s inheriting from an `Inherited` parent is:

```dart
@mustCallSuper
void deactivate() {
  assert(_lifecycleState == _ElementLifecycle.active);
  assert(_widget != null);
  assert(depth != null);
  if (_dependencies != null && _dependencies!.isNotEmpty) {
    for (final InheritedElement dependency in _dependencies!) {
      dependency._dependents.remove(this); // ... why not `dependency.removeDependent(this)`?
    }

  _inheritedWidgets = null;
  _lifecycleState = _ElementLifecycle.inactive;
}
```

Given the overridable `setDependencies`, `updateDependencies` (called by `dependOnInheritedElement`, that is called by `dependOnInheritedWidgetOfExactType` of the child itself) and `notifyDependent` that, when overridden incorrectly, can completely break the logic of inheritance on their own - this makes no sense at all. All benchmarks aside, this just looks an oversight.

As a temporary patch in my own project, I only had to do:

```dart
dependency.._dependents.remove(this)..onRemovedDependent(this); // in `deactivate()` of Element
```

With an additional empty declaration:

```dart
void onRemovedDependent(Element dependent) {} // in the body of the `InheritedElement` class
```

Two lines of code. That's all it takes.

If it's an absolute "must" that no one should ever be able accidentally override a much more imperative-style `removeDependent(Element element)` with a logic of their own that wouldn't clean-up the deactivated `Element` from the private `_dependents` field - an `onRemovedDependent` can be added there in its place just as fine. I wouldn't know why someone would want to do that, as long as the method is marked `@mustCalledSuper`, but that's just as viable of an option.

What's not viable is leaving a crucial part of cleanup up to a completely separate dependent `Element` itself.

**Comment by techvx** [comment]
Just so we're perfectly clear on the current state of affairs. 

In the body of `InheritedElement`, there are:

<details><summary>updateDependencies</summary>
<p>

```dart
@protected
void updateDependencies(Element dependent, Object? aspect) {
  setDependencies(dependent, null);
}
```

</p>
</details> 


<details><summary>getDependencies</summary>
<p>

```dart
@protected
Object? getDependencies(Element dependent) {
  return _dependents[dependent];
}
```

</p>
</details> 

<details><summary>setDependencies</summary>
<p>

```dart
@protected
void setDependencies(Element dependent, Object? value) {
  _dependents[dependent] = value;
}
```

</p>
</details> 

<details><summary>notifyDependent</summary>
<p>

```dart
@protected
void notifyDependent(covariant InheritedWidget oldWidget, Element dependent) {
  dependent.didChangeDependencies();
}
```

</p>
</details> 

<details><summary>notifyClients</summary>
<p>

```dart
@override
void notifyClients(InheritedWidget oldWidget) {
  assert(_debugCheckOwnerBuildTargetExists('notifyClients'));
  for (final Element dependent in _dependents.keys) {
    assert(() {
      Element? ancestor = dependent._parent;
      while (ancestor != this && ancestor != null) {
        ancestor = ancestor._parent;
      }
      return ancestor == this;
    }());
    assert(dependent._dependencies!.contains(this));
    notifyDependent(oldWidget, dependent); // key
  }
}
```

</p>
</details> 

Where is `removeDependent`? And why is its logic outsourced to the `deactivate()` of the `Element`?

<details><summary>deactivate</summary>
<p>

```dart
@mustCallSuper
void deactivate() {
  assert(_lifecycleState == _ElementLifecycle.active);
  assert(_widget != null);
  assert(depth != null);
  if (_dependencies != null && _dependencies!.isNotEmpty) {
    for (final InheritedElement dependency in _dependencies!) {
      dependency._dependents.remove(this); // WHY?
    }
  }
  _inheritedWidgets = null;
  _lifecycleState = _ElementLifecycle.inactive;
}
```

</p>
</details>

**Comment by rrousselGit** [comment]
The argument @Hixie made against `removeDependency` was that it removes a polymorphic call when dependencies are removed, thanks to the inlining of the logic.

So if that's the argument against `removeDependency`, then a benchmark is needed.

**Comment by techvx** [comment]
With `notifyClients` blindly iterating through the whole list of dependents without ever giving an opportunity for a custom quicker lookup and match when changes are bubbled from the parent of `InheritedElement`, this feels way off anything even remotely optimized for performance.

Even in the case of Riverpod - wouldn't it make things a whole lot more simple if you could send down the widget tree only the providers that were changed, then quickly look up the elements listening to those providers, and only them? But you can't, because someone decided to inline the logic for that - forcing you into reimplementing both the `Inherited` parent, and the `ConsumerWidget`-s below.

Not buying into the argument. If the status quo is fine, so be it - but don't tell me it's all about performance and speed when there aren't even proper benchmarks around.

**Comment by Hixie** [comment]
I'm not sure what problem we're solving here. Can you describe the code you're trying to write and why it currently fails because of the lack of the API you are proposing?

In general I would recommend always writing an issue with a fresh slate that explains things from first principles, and then only linking to additional context in later comments, because otherwise it's very confusing for people without context to understand what is happening. This issue, for example, is almost all just arguing against previously stated opinions from people whose opinions may have changed, may have stopped working on the project, or may have just forgotten all the context from those discussions, but it never actually states a concrete problem.

**Comment by techvx** [comment]
Sure: a `StatefulWidget` immediately above the `Inherited` one that manages and tracks application's state - cleaning it up, as necessary. It is important to keep it relatively local and centralized, to avoid additional hurdles with type-based inheritance that is currently relied on for the `Element` classes via their respective `_inheritedWidgets` map of `<Type, InheritedElement>`.

When it propagates the changes downward, some of them are going to be relatively minor - affecting only a handful of dependent widgets down the tree. An `InheritedModel` seemed like a great starting point, as it has, in addition to the `updateShouldNotify`, the `updateShouldNotifyDependent` method that keeps track of particular "aspects" of the state to only rebuild when necessary. Purely relying on it, however, means running the `notifyClients` loop over all dependents, every time:

```dart
for (final Element dependent in _dependents.keys) {
  // ...
  notifyDependent(oldWidget, dependent);
}
```

For relatively complex aspects, that might require expensive recalculations. I'd expect to have an opportunity to be able to link particular "aspects" to specific elements themselves - this way, when a change is made in the `StatefulWidget` above, it can be recorded on its own, than matched by aspect to the elements that listen to it - and `didChangeDependencies()` them only.

Unfortunately, the `_dependents` field is private, and with current `InheritedElement` array of methods, I can only `set`, `get`, `update` and `notify` dependents - but never clean them up. Once I link a particular `Element` to an aspect of its own, the only way to clean after it, given the performance optimized `dependency._dependents.remove(this)` within the `deactivate()` of the `Element`, is to go through the whole stack of links one-by-one and check whether any `Element` has become unmounted or not.

In addition to it being "ugly", it can get out of hand quite quick for a large set of dependents.

There's no "removeDependent" method to override. The `_dependents` are not accessible, and can't simply be replaced with a `ValueNotifier` that would listen to each and every `remove` called on it. Linking up the `Elements` in the `InheritedWidget`'s elements by necessity also requires custom overriding for the `Element` of consuming widgets down below - mainly for the `deactivate()` part, in my case. And I would strongly prefer to avoid creating yet another `ConsumerWidget` of my own to replace each and every `Stateless` widget I'd otherwise use because of one hard-coded line.

It very well be that I have misread the situation here, guys - I understand that there may very well be exceptions to a lot of commonplace rules and principles, including the ones that suggest one should keep the cleaning up logic of the parent class to the parent class itself, and not outsource it down to the dependent classes below. To an outside observer, which I am, it genuinely seemed like an oversight - as in definition of `ValueNotifier` that expects its listeners to clean up after themselves, and prohibits itself from doing as much. Performance considerations aside, this doesn't seem optimal.

If you believe that it is, that's fine as well - adding two lines of code for me wasn't much of an issue. But I'd imagine this problem would come back to haunt whoever will need to drill into their own custom `Inherited` implementation again and again. 

Should the polymorphic call turn out to be too expensive indeed (would we need to define our terms for just how much expense we'd be willing to tolerate, by the way?) - adding a 

```dart
scheduleMicrotask(() { // or a Future of any kind
  onDependentRemoved(this);
})
```

could work. 

_Any_ notification of the `Inherited` parent itself would work. 

I don't have issues with the hard-coded optimized part - I have an issue with the fact that it completely removes an essential bit of access to the `_dependents` map of the parent and gives it away to a child somewhere down below.

**Comment by Hixie** [comment]
Do you have an example app I could look at that demonstrates the performance problem that arises when using a regular InheritedWidget, or an InheritedModel, without cleanup?

(I would expect most dependencies to continue for the lifetime of the dependent widget... usually when the dependency changes, I would expect it to coincide with the widget being replaced by a new one, which makes the cleanup moot.)

Also, can you elaborate on how you would know which dependents to remove, and when you would remove them?

**Comment by techvx** [comment]
I'll try to scramble something together when I get a spare hour or two, shouldn't be too hard.

Just to make perfectly sure we're on the same page, from the example above:

1. Assuming there's a `StatefulWidget` that keeps track of all the state, "globally" (for now, auto dispose logic would come in later), throughout the life time of the Widget tree 

2. With an `InheritedWidget` / `InheritedModel` right below it, which is being passed on the state to keep track of, along with the corresponding logic to propagate the changes to the `Stateful` one right above it

3. Whenever any child down the tree - `Stateless` or otherwise, calls upon `dependOnInheritedWidget` / `...Element`, passing in the state "aspect" to watch and filter updates for

4. A "link" (via a `HashMap` or an alternative) would be created in the `InheritedElement` of our `Inherited` parent above, mapping the aspect being watched to the `Element` of the widget that requested for it. This one would, unfortunately, need to be kept separate and in addition to `_dependents` - because of the inherent hidden and private nature of that field.

5. Whenever a child requests for a change to happen, propagated through the `Inherited` into the `Stateful`, with it updating its state and passing it down to the `Inherited` once again - some of this state could be watched and mapped directly into corresponding "aspects" to trigger rebuilds for.

    In my own project right now the `Stateful` keeps track of several more sub-state classes, and whenever they get updated, it sends them down the `Inherited` chain via a rebuild, clearly indicating which of them, if any, got updated. Each sub-class has a corresponding "aspect" that child widgets listen on.

6. At this point, when the aspect matches with an update, the child widget's element must get an `didChangeDependencies()` notification. The `notifyClients` has been hard-coded on its own to go through the entire list of dependents one-by-one, calling `updateShouldNotifyDependent` for each of them with a corresponding aspect. 

    Not a problem for a relatively low amount of dependents and individual state parts, sure. Yet as things begin to grow, the complexity of iterating through N dependents, each of which might be watching for K aspects, that would need to be matched against the M amount of updates, quickly adds up.

7. The "link" from earlier reverses the iteration on its head. Instead of iterating over the dependents themselves, we're iterating over the state changes received from the `Stateful` parent above, and the "aspects" that show an interest in them. If there are no widgets and corresponding elements listening to them, no computing power is wasted. If there are, only the `dependents` that could be interested an update enter the loop.

8. This, however, requires additional book-keeping on every element and its aspect when the `updateDependencies` first get called. Not having an ability to reimplement the private `_dependents` logic on its own, we need to store both the `Element` in both `_dependents` and the new `HashMap` / whatever we're using to link things up. And things would be just fine at this point.

9. If only there was a way to know when an `Element` is no longer interested in any updates, perhaps after its `deactivate`-ion happened - which conveniently cleans up the private `_dependents`, but leaves the rest of the logic that might have been following its dependency along (our `updateDependencies`, for instance) hanging - and completely unaware of the fact that the `Element` it is supposed to notify is defunct, and "will never get mounted again" - if I remember that one correctly.

10. At this point, at best - the next propagated update from `Stateful` to `Inherited` will trigger a sub-classed aspect update via our link map from earlier, and the `InheritedElement` itself will detect that the aspect-listening `Element` has been unmounted - something it wouldn't even have to check for if there was a way for it to get notified of the whole deactivation / unmountedness earlier. At worst, if the same kind of an update never occurs again - for any reason, we have a memory leak.

11. To prevent that one, now we have to schedule regular cleanups. Each and every update would have to iterate through the whole list of elements once more and they are all still mounted or not. Processor cycles wasted, battery draining faster, possibly no changes introduced whatsoever - if elements remain the same, yet each and every minor change will trigger the re-check after re-check after re-check just to make sure we're not storing a bunch of invalid references.

12. And having quite a low tolerance for such possibilities makes you hard-code `onDependencyRemoved` into the `deactivate` yourself - alongside `_dependencies.remove(this)`.

Hopefully that explains it better. I'll send in an example when I have the time.

This is not as much about performance (for me), however, as it is about what makes and doesn't make sense. Custom inheritance makes sense. Filtered updates make sense. Iterating over `_dependents` when we can be iterating over the updates only some of them might be listening to - doesn't make sense. And having access to `updateDependencies` and custom tracking logic, but only up until the point of cleanup, unfortunately, doesn't make any sense either.

**Comment by Hixie** [comment]
This sounds quite convoluted. I look forward to seeing a concrete example of the problem you want to solve.

**Comment by s0nerik** [comment]
@Hixie @goderbauer 

Since #129207 was closed as a duplicate of this and the conversation here hints that a compelling argument for actually having this API is needed, I've made my WIP observable_values/state_management library (which requires API like this and relies on #129210 so far) public. [Here is](https://github.com/s0nerik/easy_observable/blob/ef4b360fdbe043439a8b6d729623ca5e6b1ad239/lib/src/inherited_observable_notifier.dart#L92) how my library uses the "dependent removal notification" API.

And now to the fun part. This API allows me to make observing a value within a Widget as simple as:
```dart
class _Counter extends StatelessWidget {
  const _Counter({
    Key? key,
    required this.counter,
  }) : super(key: key);

  final Observable<int> counter;

  @override
  Widget build(BuildContext context) {
    return Text('Counter: ${counter.watch(context)}');
  }
}
```

Notice how I don't need to change a base type of `_Counter` to anything else from an ordinary `StatelessWidget` to make it react to `counter` changes. Also, notice how there's no `dispose`/`close`/etc. calls anywhere here - the reason is that `InheritedWidget` machinery is used to notify about the value changes. Basically, there's a global `ObservableRoot` inherited widget that is responsible for managing all the observable value subscriptions for every element/context that defines the dependency, and `counter.watch(context)` does an "ordinary" `context.dependOnInheritedWidgetOfExactType<_InheritedObservableNotifier>(aspect: observable)` under the hood. The only major problem with this approach is the inability to detect when the dependent element is deactivated within the `ObservableRoot` so that it could clean up all the value change observers attached to that element. This is why my library needs that API.

Most of the popular state management solutions I've seen out there propose observing values by either
- changing the base type to `StatelessObserverWidget`/`ConsumerWidget`/etc., which doesn't play well with `const` constructors in hot reload scenarios.
- wrapping the widget in `ObserverBuilder`/`Consumer`/etc., which makes it less convenient to use.

I suspect that the lack of "element deactivation notification" API might have been one of the reasons existing state management solutions went that way.

The approach I've presented above is the sweet spot IMO, as it plays well with hot reload even with `const` constructors (which is VERY common in practice), and looks quite natural and easy to use to me.

I can even imagine this approach available out of the box for `ValueNotifier`s, as there seems to be nothing (except for this ticket's API) that conceptually stops us from having it. I can try doing a PR for that if there's any interest in having this in-built. Meanwhile, I'll definitely look into adding the same `.watch(context)` capability to ordinary `ValueNotifier`s in my library (maybe even making the `Observable`s be `ValueNotifier`s themselves).

**Comment by goderbauer** [comment]
@s0nerik Could you extend your counter example from above a little bit to guide with understanding? Where/how does the value for count get injected into the tree? 

From the example given it is not clear to my why you would need additional API. I would expect that "watch" creates a dependency between whatever injects the counter into the tree and the "_Counter". The normal dependency machinery would inform the "Counter" widget to rebuild if whatever injects the counter value gets a new value. And that dependency would cease to exist if "Counter" is deactivated. Why are additional actions needed on deactivation?

**Comment by s0nerik** [comment]
@goderbauer

### TL;DR;

> Why are additional actions needed on deactivation?

(1) To ensure that dependent `Element` and `Observable` references are removed from `ObservableRoot`(`_InheritedObservableNotifier`) when the dependent `Element` is deactivated.
(2) To ensure that all `Observable` subscriptions are properly disposed of when the dependent `Element` is deactivated.

### Explanation

> I would expect that "watch" creates a dependency between whatever injects the counter into the tree and the "_Counter".

This is not the case here. The dependency is created between the `_Counter` and `ObservableRoot` ([`_InheritedObservableNotifier`](https://github.com/s0nerik/easy_observable/blob/ef4b360fdbe043439a8b6d729623ca5e6b1ad239/lib/src/inherited_observable_notifier.dart#L24) underneath it, to be precise). `ObservableRoot` is a global manager for all the `Observable` subscriptions made via `watch` and is the one that calls `markNeedsBuild` on the proper `Element`s. It is meant to be put somewhere at the top of the widgets tree, above the `MaterialApp`, like this:

```dart
ObservableRoot(
  child: MaterialApp(
    ...
      _Counter()
      _OtherObserver()
  ),
);
```

It needs to be an inherited widget for (1) quick access and (2) the ability of `InheritedElement` to receive a callback with both the `Observable` that needs to be observed and the `Element` interested in observing it.

> The normal dependency machinery would inform the "Counter" widget to rebuild if whatever injects the counter value gets a new value.

Here's how `_InheritedObservableNotifier` is defined:
```dart
class _InheritedObservableNotifier extends InheritedWidget {
  const _InheritedObservableNotifier({super.key, required super.child});

  @override
  InheritedElement createElement() => _ObservableNotifierInheritedElement(this);

  @override
  bool updateShouldNotify(covariant InheritedWidget oldWidget) => false;
}
```

You may notice that `_InheritedObservableNotifier` never actually notifies of any changes via `updateShouldNotify()` as "normal" inherited widgets do. Instead, it creates a custom `_ObservableNotifierInheritedElement` which is in charge of creating/updating/disposing `StreamSubscription`s for every `Observable` field being `watch`ed. Creating/updating the `StreamSubscription`s is done within the `updateDependencies()` method of that element.

Disposal of `StreamSubscriptions` for no-longer-referenced `Observable`s is the trickiest part and is done in a `postFrameCallback` by comparing the set of `Observable`s referenced within an `Element` in the last frame with those that were referenced before:

```dart
  void _onPostFrame(_) {
    if (!mounted) return;

    // dispose all subscriptions that are no longer present in the frame
    // subscriptions map, but only for elements that are present within the
    // frame element subscriptions map
    for (final element in _frameElementSubs.keys) {
      final frameObservableSubs = _frameElementSubs[element]!;
      final observableSubs = _elementSubs[element]!;
      for (final observable in observableSubs.keys) {
        if (!frameObservableSubs.containsKey(observable)) {
          final sub = observableSubs[observable]!;
          sub.cancel();
        }
      }
      _elementSubs[element] = frameObservableSubs;
      if (observableSubs.isEmpty) {
        _elementSubs.remove(element);
      }
    }

    _frameElementSubs.clear();
    SchedulerBinding.instance.addPostFrameCallback(_onPostFrame);
  }
```

The code above works well for disposing of no-longer-referenced `Observable` subscriptions for an `Element` when going from `watch()`ing 2+ `Observable`s to 1, but is useless when:
- `watch()`es disappear from `build` completely

  I already have a quite easy-to-use [workaround](https://github.com/s0nerik/easy_observable/blob/ef4b360fdbe043439a8b6d729623ca5e6b1ad239/lib/src/inherited_observable_notifier.dart#L125) for handling that case, but it needs to be applied at the use-site, and I consider not applying it mostly harmless, as it only helps to avoid some extra rebuilds for alive widgets (which is usually not really a big problem, IMO). It solves the following cases:
  - `watch()` stops being called within the the `build` due to it being used conditionally
  - all `watch()`es are removed from the `build` in code and hot reload happens
  
  This workaround shouldn't be needed if https://github.com/flutter/flutter/issues/106549 gets implemented.

- the `Element` gets deactivated
  
  **This is the root problem.**

  > Why are additional actions needed on deactivation?
  
  Because `_ObservableNotifierInheritedElement` needs to dispose of all `StreamSubscriptions` for a dependent `Element` that was deactivated.
  
  With `InheritedElement.removeDependent(Element element)` it would be as easy as
  ```dart
  void removeDependent(Element element) {
    super.removeDependent(element);
    _dispose(element);
  }
  ```
  
> Where/how does the value for count get injected into the tree?

It doesn't really matter for the example, `Observable` value can be provided from anywhere (global/top-level variable, `Provider`/InheritedWidget/parent Context, a property defined on a `State`, a parent Widget, etc.). As I explained above, observing happens inside the `_ObservableNotifierInheritedElement`.

**Comment by s0nerik** [comment]
I have found a [workaround](https://github.com/s0nerik/easy_observable/commit/f5bbad46cb177e0c8a181982714838893e8c2cd8) to get what I needed working, but it requires iterating over each and every `Element` that `watch()`es any observables every frame. It doesn't look like a huge performance hit to me, but is certainly wasteful when it could've been as simple as 

```dart
@override
void removeDependent(Element element) {
  super.removeDependent(element);
  _dispose(element);
}
```

While I can somewhat understand the fear of potentially downgrading performance by eagerly clearing the dependencies to handle https://github.com/flutter/flutter/issues/106549, I see no such potential here as elements are not `deactivated` that often, while having the ability to know that a dependent has been removed inside your own inherited widget is be beneficial at least in the cases mentioned here (though I'm sure there're much more potential usages to this).

**Comment by s0nerik** [comment]
Done in https://github.com/flutter/flutter/pull/129210
